bottlecap/src/traces/propagation/mod.rs

```diff
@@ -1,3 +1,873 @@
 pub mod error;
 pub mod text_map_propagator;
+
+pub trait Propagator {
+    fn extract(&self, carrier: &dyn Extractor) -> Option<SpanContext>;
+    fn inject(&self, context: SpanContext, carrier: &mut dyn Injector);
+}
+
+pub struct DatadogCompositePropagator {
+    propagators: Vec<Box<dyn Propagator + 'static>>,
+    config: Arc<config::Config>,
+}
+
+#[allow(clippy::never_loop)]
+impl Propagator for DatadogCompositePropagator {
+    fn extract(&self, carrier: &dyn Extractor) -> Option<SpanContext> {
+        if self.config.trace_propagation_extract_first {
+            for propagator in &self.propagators {
+                let context = propagator.extract(carrier);
+
+                if self.config.trace_propagation_http_baggage_enabled {
+                    if let Some(mut context) = context {
+                        Self::attach_baggage(&mut context, carrier);
+                        return Some(context);
+                    }
+                }
+
+                return context;
+            }
+        }
+
+        let (contexts, styles) = self.extract_available_contexts(carrier);
+        if contexts.is_empty() {
+            return None;
+        }
+
+        let mut context = Self::resolve_contexts(contexts, styles, carrier);
+        if self.config.trace_propagation_http_baggage_enabled {
+            Self::attach_baggage(&mut context, carrier);
+        }
+
+        Some(context)
+    }
+
+    fn inject(&self, _context: SpanContext, _carrier: &mut dyn Injector) {
+        todo!()
+    }
+}
+
+impl DatadogCompositePropagator {
+    #[must_use]
+    pub fn new(config: Arc<config::Config>) -> Self {
+        let propagators: Vec<Box<dyn Propagator + 'static>> = config
+            .trace_propagation_style_extract
+            .iter()
+            .filter_map(|style| match style {
+                TracePropagationStyle::Datadog => {
+                    Some(Box::new(text_map_propagator::DatadogHeaderPropagator)
+                        as Box<dyn Propagator>)
+                }
+                TracePropagationStyle::TraceContext => {
+                    Some(Box::new(text_map_propagator::TraceContextPropagator)
+                        as Box<dyn Propagator>)
+                }
+                _ => None,
+            })
+            .collect();
+
+        Self {
+            propagators,
+            config,
+        }
+    }
+
+    fn extract_available_contexts(
+        &self,
+        carrier: &dyn Extractor,
+    ) -> (Vec<SpanContext>, Vec<TracePropagationStyle>) {
+        let mut contexts = Vec::<SpanContext>::new();
+        let mut styles = Vec::<TracePropagationStyle>::new();
+
+        for (i, propagator) in self.propagators.iter().enumerate() {
+            if let Some(context) = propagator.extract(carrier) {
+                contexts.push(context);
+                styles.push(self.config.trace_propagation_style_extract[i]);
+            }
+        }
+
+        (contexts, styles)
+    }
+
+    fn resolve_contexts(
```

Comment:
can this method be outside of the struct implementation?

Reply:
Since this is the only place being used, I'd rather keep it here, there's no other place where this is being used. What would be the benefit of keeping it outside of the struct?